monai/apps/auto3dseg/ensemble_builder.py

```diff
@@ -0,0 +1,251 @@
+#     http://www.apache.org/licenses/LICENSE-2.0
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+import json
+import logging
+import os
+from abc import ABC, abstractmethod
+from copy import deepcopy
+from typing import Any, Dict, List, Optional, Sequence
+
+import numpy as np
+
+from monai.apps.auto3dseg.bundle_gen import BundleAlgo
+from monai.auto3dseg import concat_val_to_np
+from monai.bundle import ConfigParser
+from monai.utils.enums import BundleEnsembleKeys
+
+logger = logging.getLogger(__name__)
+
+
+class BundleEnsemble(ABC):
+    """
+    The base class of Ensemble methods
+    """
+
+    def __init__(self):
+        self.algos = []
+        self.infer_files = []
+        self.algo_ensemble = []
+
+    def set_algos(self, infer_algos):
+        """
+        Register model in the ensemble
+        """
+        self.algos = deepcopy(infer_algos)
+
+    def get_algo(self, identifier):
+        """
+        Get a model by identifier.
+
+        Args:
+            identifier: the name of the bundleAlgo
+        """
+        for algo in self.algos:
+            if identifier == algo[BundleEnsembleKeys.ID]:
+                return algo
+
+    def get_algo_ensemble(self):
+        """
+        Get the algo ensemble after ranking or a empty list if ranking was not started.
+
+        Returns:
+            A list of Algo
+        """
+        return self.algo_ensemble
+
+    def set_infer_files(self, dataroot: str, data_src_cfg_file: str):
+        """
+        Set the files to perform model inference.
+
+        Args:
+            dataroot: the path of the files
+            data_src_cfg_file: the data source file path
+        """
+        with open(data_src_cfg_file) as f:
+            datalist = json.load(f)
+
+        for d in datalist["testing"]:
+            self.infer_files.append({"image": os.path.join(dataroot, d["image"])})
+
+    @abstractmethod
+    def rank_algos(self):
+        raise NotImplementedError
+
+    @abstractmethod
+    def predict(self):
+        """
+        predict results after the models are ranked/weighted
+        """
+        raise NotImplementedError
+
+
+class BundleEnsembleBestN(BundleEnsemble):
+    """
+    Ensemble method that select N model out of all using the models' best_metric scores
+
+    Args:
+        n_best: number of models to pick for ensemble (N).
+    """
+
+    def __init__(self, n_best: int = 5):
+
+        super().__init__()
+        self.n_best = n_best
+
+    def sort_score(self):
+        """
+        Sort the best_metrics
+        """
+        scores = concat_val_to_np(self.algos, [BundleEnsembleKeys.SCORE])
+        return np.argsort(scores).tolist()
+
+    def rank_algos(self):
```

Comment:
we should have an additional method for ranking in each individual folds. we chose best model for each cross-validation fold in previous public challenges.

Reply:
Added a new class to address the need. Thanks for the suggestion!

```python
class BundleEnsembleBestByFold(BundleEnsemble):
    """
    Ensemble method that select the best models that are the tops in each fold.

    Args:
        n_fold: number of cross-valiation folds used in training
    """

    def __init__(self, n_fold: int = 5):
```